File: app/src/main/java/com/drdisagree/iconify/ui/NavigationBar.java
package com.drdisagree.iconify.ui;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.Switch;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.drdisagree.iconify.Iconify;
import com.drdisagree.iconify.R;
import com.drdisagree.iconify.config.PrefConfig;
import com.drdisagree.iconify.utils.OverlayUtils;
import com.google.android.material.appbar.CollapsingToolbarLayout;
import com.topjohnwu.superuser.Shell;

import java.util.List;
import java.util.Objects;

public class NavigationBar extends AppCompatActivity {

    private ViewGroup container;
    List<String> left_back_gesture = Shell.cmd("settings get secure back_gesture_inset_scale_left").exec().getOut();
    List<String> right_back_gesture = Shell.cmd("settings get secure back_gesture_inset_scale_right").exec().getOut();

    // Switches
    @SuppressLint("UseSwitchCompatOrMaterialCode")
    Switch nb_fullscreen, nb_immersive, nb_immersivev2, nb_immersivev3, nb_hide_pill, nb_monet_pill, nb_disable_left_gesture, nb_disable_right_gesture, nb_lower_sens, nb_hide_kb_buttons;

    // Views
    LinearLayout nb_pill_menu, nb_monet_pill_menu, nb_kb_buttons_menu;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_navigation_bar);

        // Header
        CollapsingToolbarLayout collapsing_toolbar = findViewById(R.id.collapsing_toolbar);
        collapsing_toolbar.setTitle(getResources().getString(R.string.activity_title_navigation_bar));
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);

        // Switches
        nb_fullscreen = findViewById(R.id.nb_fullscreen);
        nb_immersive = findViewById(R.id.nb_immersive);
        nb_immersivev2 = findViewById(R.id.nb_immersivev2);
        nb_immersivev3 = findViewById(R.id.nb_immersivev3);
        nb_lower_sens = findViewById(R.id.nb_lower_sens);
        nb_hide_pill = findViewById(R.id.nb_hide_pill);
        nb_monet_pill = findViewById(R.id.nb_monet_pill);
        nb_hide_kb_buttons = findViewById(R.id.nb_hide_kb_buttons);
        nb_disable_left_gesture = findViewById(R.id.nb_disable_left_gesture);
        nb_disable_right_gesture = findViewById(R.id.nb_disable_right_gesture);

        // Views
        nb_pill_menu = findViewById(R.id.nb_pill_menu);
        nb_monet_pill_menu = findViewById(R.id.nb_monet_pill_menu);
        nb_kb_buttons_menu = findViewById(R.id.nb_kb_buttons_menu);

        // Switch states
        nb_fullscreen.setChecked(PrefConfig.loadPrefBool(Iconify.getAppContext(), "IconifyComponentNBFullScreen.overlay"));
        nb_immersive.setChecked(PrefConfig.loadPrefBool(Iconify.getAppContext(), "IconifyComponentNBImmersive.overlay"));
        nb_immersivev2.setChecked(PrefConfig.loadPrefBool(Iconify.getAppContext(), "IconifyComponentNBImmersiveSmall.overlay"));
        nb_immersivev3.setChecked(PrefConfig.loadPrefBool(Iconify.getAppContext(), "IconifyComponentNBImmersiveSmaller.overlay"));
        nb_lower_sens.setChecked(PrefConfig.loadPrefBool(Iconify.getAppContext(), "IconifyComponentNBLowSens.overlay"));
        nb_hide_pill.setChecked(PrefConfig.loadPrefBool(Iconify.getAppContext(), "IconifyComponentNBHidePill.overlay"));
        nb_monet_pill.setChecked(PrefConfig.loadPrefBool(Iconify.getAppContext(), "IconifyComponentNBMonetPill.overlay"));
        nb_hide_kb_buttons.setChecked(PrefConfig.loadPrefBool(Iconify.getAppContext(), "IconifyComponentNBHideKBButton.overlay"));
        nb_disable_left_gesture.setChecked(initialize_left_gesture_switch());
        nb_disable_right_gesture.setChecked(initialize_right_gesture_switch());

        if (nb_immersive.isChecked() || nb_immersivev2.isChecked() || nb_immersivev3.isChecked()) {
            nb_kb_buttons_menu.setVisibility(View.VISIBLE);
        }

        if (nb_fullscreen.isChecked()) {
            nb_pill_menu.setVisibility(View.GONE);
        }

        if (nb_hide_pill.isChecked()) {
            nb_monet_pill_menu.setVisibility(View.GONE);
        }

        // Fullscreen
        nb_fullscreen.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                nb_pill_menu.setVisibility(View.GONE);
                nb_kb_buttons_menu.setVisibility(View.VISIBLE);
                disableOthers("IconifyComponentNBFullScreen.overlay");
                OverlayUtils.enableOverlay("IconifyComponentNBFullScreen.overlay");
            } else {
                nb_pill_menu.setVisibility(View.VISIBLE);
                nb_kb_buttons_menu.setVisibility(View.GONE);
                OverlayUtils.disableOverlay("IconifyComponentNBFullScreen.overlay");
            }
        });

        // Immersive
        nb_immersive.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                nb_kb_buttons_menu.setVisibility(View.VISIBLE);
                disableOthers("IconifyComponentNBImmersive.overlay");
                OverlayUtils.enableOverlay("IconifyComponentNBImmersive.overlay");
            } else {
                nb_kb_buttons_menu.setVisibility(View.GONE);
                OverlayUtils.disableOverlay("IconifyComponentNBImmersive.overlay");
            }
        });

        // Immersive v2
        nb_immersivev2.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                nb_kb_buttons_menu.setVisibility(View.VISIBLE);
                disableOthers("IconifyComponentNBImmersiveSmall.overlay");
                OverlayUtils.enableOverlay("IconifyComponentNBImmersiveSmall.overlay");
            } else {
                nb_kb_buttons_menu.setVisibility(View.GONE);
                OverlayUtils.disableOverlay("IconifyComponentNBImmersiveSmall.overlay");
            }
        });

        // Immersive v3
        nb_immersivev3.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                nb_kb_buttons_menu.setVisibility(View.VISIBLE);
                disableOthers("IconifyComponentNBImmersiveSmaller.overlay");
                OverlayUtils.enableOverlay("IconifyComponentNBImmersiveSmaller.overlay");
            } else {
                nb_kb_buttons_menu.setVisibility(View.GONE);
                OverlayUtils.disableOverlay("IconifyComponentNBImmersiveSmaller.overlay");
            }
        });

        // Lower Sensitivity
        nb_hide_pill.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                OverlayUtils.enableOverlay("IconifyComponentNBLowSens.overlay");
            } else {
                OverlayUtils.disableOverlay("IconifyComponentNBLowSens.overlay");
            }
        });

        // Hide Pill
        nb_hide_pill.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                nb_monet_pill_menu.setVisibility(View.GONE);
                OverlayUtils.enableOverlay("IconifyComponentNBHidePill.overlay");
                Shell.cmd("killall com.android.systemui").exec();
            } else {
                nb_monet_pill_menu.setVisibility(View.VISIBLE);
                OverlayUtils.disableOverlay("IconifyComponentNBHidePill.overlay");
                Shell.cmd("killall com.android.systemui").exec();
            }
        });

        // Monet Pill
        nb_monet_pill.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                OverlayUtils.enableOverlay("IconifyComponentNBMonetPill.overlay");
                Shell.cmd("killall com.android.systemui").exec();
            } else {
                OverlayUtils.disableOverlay("IconifyComponentNBMonetPill.overlay");
                Shell.cmd("killall com.android.systemui").exec();
            }
        });

        // Hide Keyboard Buttons
        nb_hide_kb_buttons.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                OverlayUtils.enableOverlay("IconifyComponentNBHideKBButton.overlay");
            } else {
                OverlayUtils.disableOverlay("IconifyComponentNBHideKBButton.overlay");
            }
        });

        // Disable left gesture
        nb_disable_left_gesture.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                Shell.cmd("settings put secure back_gesture_inset_scale_left -1 &>/dev/null").exec();
            } else {
                Shell.cmd("settings delete secure back_gesture_inset_scale_left &>/dev/null").exec();
            }
        });

        // Disable right gesture
        nb_disable_right_gesture.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                Shell.cmd("settings put secure back_gesture_inset_scale_right -1 &>/dev/null").exec();
            } else {
                Shell.cmd("settings delete secure back_gesture_inset_scale_right &>/dev/null").exec();
            }
        });
    }

    private boolean initialize_left_gesture_switch() {
        boolean left_gesture = false;
        try {
            left_gesture = Integer.parseInt(left_back_gesture.get(0)) == -1;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return left_gesture;
    }

    private boolean initialize_right_gesture_switch() {
        boolean right_gesture = false;
        try {
            right_gesture = Integer.parseInt(right_back_gesture.get(0)) == -1;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return right_gesture;
    }

    private void disableOthers(String pkgName) {
        if (!Objects.equals(pkgName, "IconifyComponentNBFullScreen.overlay")) {
            nb_fullscreen.setChecked(false);
            PrefConfig.savePrefBool(Iconify.getAppContext(), ("IconifyComponentNBFullScreen.overlay"), false);
            OverlayUtils.disableOverlay("IconifyComponentNBFullScreen.overlay");
        }
        if (!Objects.equals(pkgName, "IconifyComponentNBImmersive.overlay")) {
            nb_immersive.setChecked(false);
            PrefConfig.savePrefBool(Iconify.getAppContext(), ("IconifyComponentNBImmersive.overlay"), false);
            OverlayUtils.disableOverlay("IconifyComponentNBImmersive.overlay");
        }
        if (!Objects.equals(pkgName, "IconifyComponentNBImmersiveSmall.overlay")) {
            nb_immersivev2.setChecked(false);
            PrefConfig.savePrefBool(Iconify.getAppContext(), ("IconifyComponentNBImmersiveSmall.overlay"), false);
            OverlayUtils.disableOverlay("IconifyComponentNBImmersiveSmall.overlay");
        }
        if (!Objects.equals(pkgName, "IconifyComponentNBImmersiveSmaller.overlay")) {
            nb_immersivev3.setChecked(false);
            PrefConfig.savePrefBool(Iconify.getAppContext(), ("IconifyComponentNBImmersiveSmaller.overlay"), false);
            OverlayUtils.disableOverlay("IconifyComponentNBImmersiveSmaller.overlay");
        }
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }
}

File: app/src/main/java/com/drdisagree/iconify/ui/fragments/tweaks/NavigationBar.kt
package com.drdisagree.iconify.ui.fragments.tweaks

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CompoundButton
import androidx.annotation.IntRange
import com.drdisagree.iconify.R
import com.drdisagree.iconify.common.Const.FRAMEWORK_PACKAGE
import com.drdisagree.iconify.common.Const.SWITCH_ANIMATION_DELAY
import com.drdisagree.iconify.common.Const.SYSTEMUI_PACKAGE
import com.drdisagree.iconify.common.Preferences.NAVBAR_FULL_SCREEN
import com.drdisagree.iconify.common.Preferences.NAVBAR_GCAM_LAG_FIX
import com.drdisagree.iconify.common.Preferences.NAVBAR_HIDE_PILL
import com.drdisagree.iconify.common.Preferences.NAVBAR_IMMERSIVE_V1
import com.drdisagree.iconify.common.Preferences.NAVBAR_IMMERSIVE_V2
import com.drdisagree.iconify.common.Preferences.NAVBAR_IMMERSIVE_V3
import com.drdisagree.iconify.common.Preferences.NAVBAR_LOW_SENS
import com.drdisagree.iconify.common.Preferences.PILL_SHAPE_SWITCH
import com.drdisagree.iconify.common.References.FABRICATED_PILL_BOTTOM_SPACE
import com.drdisagree.iconify.common.References.FABRICATED_PILL_THICKNESS
import com.drdisagree.iconify.common.References.FABRICATED_PILL_WIDTH
import com.drdisagree.iconify.config.RPrefs.getBoolean
import com.drdisagree.iconify.config.RPrefs.getInt
import com.drdisagree.iconify.config.RPrefs.putBoolean
import com.drdisagree.iconify.config.RPrefs.putInt
import com.drdisagree.iconify.databinding.FragmentNavigationBarBinding
import com.drdisagree.iconify.ui.base.BaseFragment
import com.drdisagree.iconify.ui.utils.ViewHelper.setHeader
import com.drdisagree.iconify.utils.SystemUtils
import com.drdisagree.iconify.utils.SystemUtils.hasStoragePermission
import com.drdisagree.iconify.utils.SystemUtils.requestStoragePermission
import com.drdisagree.iconify.utils.SystemUtils.restartSystemUI
import com.drdisagree.iconify.utils.overlay.OverlayUtils
import com.drdisagree.iconify.utils.overlay.OverlayUtils.enableOverlay
import com.drdisagree.iconify.utils.overlay.manager.resource.ResourceEntry
import com.drdisagree.iconify.utils.overlay.manager.resource.ResourceManager.buildOverlayWithResource
import com.drdisagree.iconify.utils.overlay.manager.resource.ResourceManager.removeResourceFromOverlay
import com.google.android.material.slider.Slider
import com.topjohnwu.superuser.Shell
import java.util.concurrent.atomic.AtomicBoolean

class NavigationBar : BaseFragment() {

    private lateinit var binding: FragmentNavigationBarBinding

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentNavigationBarBinding.inflate(inflater, container, false)
        val view: View = binding.getRoot()

        // Header
        setHeader(
            requireContext(),
            getParentFragmentManager(),
            binding.header.toolbar,
            R.string.activity_title_navigation_bar
        )

        // Switch states
        binding.nbFullscreen.isSwitchChecked = getBoolean(NAVBAR_FULL_SCREEN)
        binding.nbImmersive.isSwitchChecked = getBoolean(NAVBAR_IMMERSIVE_V1)
        binding.nbImmersiveV2.isSwitchChecked = getBoolean(NAVBAR_IMMERSIVE_V2)
        binding.nbImmersiveV3.isSwitchChecked = getBoolean(NAVBAR_IMMERSIVE_V3)
        binding.nbGcamLagFix.isSwitchChecked = getBoolean(NAVBAR_GCAM_LAG_FIX)
        binding.nbLowerSens.isSwitchChecked = getBoolean(NAVBAR_LOW_SENS)
        binding.nbHidePill.isSwitchChecked = getBoolean(NAVBAR_HIDE_PILL)
        binding.nbMonetPill.isSwitchChecked =
            getBoolean("IconifyComponentNBMonetPill.overlay")
        binding.nbHideKbButtons.isSwitchChecked =
            getBoolean("IconifyComponentNBHideKBButton.overlay")

        binding.nbDisableLeftGesture.isSwitchChecked = initializeLeftGestureSwitch()
        binding.nbDisableRightGesture.isSwitchChecked = initializeRightGestureSwitch()

        binding.nbHidePill.setEnabled(!binding.nbFullscreen.isSwitchChecked)
        binding.nbMonetPill.setEnabled(!binding.nbHidePill.isSwitchChecked && !binding.nbFullscreen.isSwitchChecked)

        // Fullscreen
        val nbFullScreenClicked = AtomicBoolean(false)

        binding.nbFullscreen.setSwitchChangeListener { buttonView: CompoundButton, isSwitchChecked: Boolean ->
            if (buttonView.isPressed || nbFullScreenClicked.get()) {
                nbFullScreenClicked.set(false)

                if (!hasStoragePermission()) {
                    requestStoragePermission(requireContext())
                    binding.nbFullscreen.isSwitchChecked = !isSwitchChecked

                    return@setSwitchChangeListener
                }

                binding.nbHidePill.setEnabled(!isSwitchChecked)
                binding.nbMonetPill.setEnabled(!isSwitchChecked && !binding.nbHidePill.isSwitchChecked)

                disableOthers(NAVBAR_FULL_SCREEN)

                if (isSwitchChecked) {
                    binding.pillShape.pillShapeContainer.visibility = View.GONE
                } else {
                    binding.pillShape.pillShapeContainer.visibility = View.VISIBLE
                }

                Handler(Looper.getMainLooper()).postDelayed(
                    { handleFullScreen(isSwitchChecked) },
                    SWITCH_ANIMATION_DELAY
                )
            }
        }
        binding.nbFullscreen.setBeforeSwitchChangeListener {
            nbFullScreenClicked.set(
                true
            )
        }

        // Immersive
        val nbImmersiveClicked = AtomicBoolean(false)

        binding.nbImmersive.setSwitchChangeListener { buttonView: CompoundButton, isSwitchChecked: Boolean ->
            if (buttonView.isPressed || nbImmersiveClicked.get()) {
                nbImmersiveClicked.set(false)

                if (!hasStoragePermission()) {
                    requestStoragePermission(requireContext())
                    binding.nbImmersive.isSwitchChecked = !isSwitchChecked

                    return@setSwitchChangeListener
                }

                disableOthers(NAVBAR_IMMERSIVE_V1)

                Handler(Looper.getMainLooper()).postDelayed(
                    { handleImmersive(isSwitchChecked, 1) },
                    SWITCH_ANIMATION_DELAY
                )
            }
        }
        binding.nbImmersive.setBeforeSwitchChangeListener {
            nbImmersiveClicked.set(
                true
            )
        }

        // Immersive V2
        val nbImmersiveV2Clicked = AtomicBoolean(false)
        binding.nbImmersiveV2.setSwitchChangeListener { buttonView: CompoundButton, isSwitchChecked: Boolean ->
            if (buttonView.isPressed || nbImmersiveV2Clicked.get()) {
                nbImmersiveV2Clicked.set(false)

                if (!hasStoragePermission()) {
                    requestStoragePermission(requireContext())
                    binding.nbImmersiveV2.isSwitchChecked = !isSwitchChecked

                    return@setSwitchChangeListener
                }

                disableOthers(NAVBAR_IMMERSIVE_V2)

                Handler(Looper.getMainLooper()).postDelayed(
                    { handleImmersive(isSwitchChecked, 2) },
                    SWITCH_ANIMATION_DELAY
                )
            }
        }
        binding.nbImmersiveV2.setBeforeSwitchChangeListener {
            nbImmersiveV2Clicked.set(
                true
            )
        }

        // Immersive V3
        val nbImmersiveV3Clicked = AtomicBoolean(false)
        binding.nbImmersiveV3.setSwitchChangeListener { buttonView: CompoundButton, isSwitchChecked: Boolean ->
            if (buttonView.isPressed || nbImmersiveV3Clicked.get()) {
                nbImmersiveV3Clicked.set(false)

                if (!hasStoragePermission()) {
                    requestStoragePermission(requireContext())
                    binding.nbImmersiveV3.isSwitchChecked = !isSwitchChecked

                    return@setSwitchChangeListener
                }

                disableOthers(NAVBAR_IMMERSIVE_V3)

                Handler(Looper.getMainLooper()).postDelayed(
                    { handleImmersive(isSwitchChecked, 3) },
                    SWITCH_ANIMATION_DELAY
                )
            }
        }
        binding.nbImmersiveV3.setBeforeSwitchChangeListener {
            nbImmersiveV3Clicked.set(
                true
            )
        }

        // GCam Lag Fix
        val nbGcamLagFixClicked = AtomicBoolean(false)
        binding.nbGcamLagFix.setSwitchChangeListener { buttonView: CompoundButton, isSwitchChecked: Boolean ->
            if (buttonView.isPressed || nbGcamLagFixClicked.get()) {
                nbGcamLagFixClicked.set(false)

                if (!hasStoragePermission()) {
                    requestStoragePermission(requireContext())
                    binding.nbGcamLagFix.isSwitchChecked = !isSwitchChecked
                    return@setSwitchChangeListener
                }

                putBoolean(NAVBAR_GCAM_LAG_FIX, isSwitchChecked)

                val fullscreen = getBoolean(NAVBAR_FULL_SCREEN)
                val immersive1 = getBoolean(NAVBAR_IMMERSIVE_V1)
                val immersive2 = getBoolean(NAVBAR_IMMERSIVE_V2)
                val immersive3 = getBoolean(NAVBAR_IMMERSIVE_V3)

                Handler(Looper.getMainLooper()).postDelayed({
                    if (fullscreen) {
                        handleFullScreen(true)
                    } else if (immersive1) {
                        handleImmersive(true, 1)
                    } else if (immersive2) {
                        handleImmersive(true, 2)
                    } else if (immersive3) {
                        handleImmersive(true, 3)
                    }
                }, SWITCH_ANIMATION_DELAY)
            }
        }
        binding.nbGcamLagFix.setBeforeSwitchChangeListener {
            nbGcamLagFixClicked.set(
                true
            )
        }

        // Lower Sensitivity
        val nbLowerSensClicked = AtomicBoolean(false)

        binding.nbLowerSens.setSwitchChangeListener { buttonView: CompoundButton, isSwitchChecked: Boolean ->
            if (buttonView.isPressed || nbLowerSensClicked.get()) {
                nbLowerSensClicked.set(false)

                if (!hasStoragePermission()) {
                    requestStoragePermission(requireContext())
                    binding.nbLowerSens.isSwitchChecked = !isSwitchChecked

                    return@setSwitchChangeListener
                }

                Handler(Looper.getMainLooper()).postDelayed(
                    { handleLowSensitivity(isSwitchChecked) },
                    SWITCH_ANIMATION_DELAY
                )
            }
        }
        binding.nbLowerSens.setBeforeSwitchChangeListener {
            nbLowerSensClicked.set(
                true
            )
        }

        // Hide Pill
        val nbHidePillClicked = AtomicBoolean(false)

        binding.nbHidePill.setSwitchChangeListener { buttonView: CompoundButton, isSwitchChecked: Boolean ->
            if (buttonView.isPressed || nbHidePillClicked.get()) {
                nbHidePillClicked.set(false)

                if (!hasStoragePermission()) {
                    requestStoragePermission(requireContext())
                    binding.nbHidePill.isSwitchChecked = !isSwitchChecked

                    return@setSwitchChangeListener
                }

                Handler(Looper.getMainLooper()).postDelayed({
                    handleHidePill(isSwitchChecked)

                    Handler(Looper.getMainLooper()).postDelayed(
                        { SystemUtils.handleSystemUIRestart() },
                        2000
                    )
                }, SWITCH_ANIMATION_DELAY)
            }
        }
        binding.nbHidePill.setBeforeSwitchChangeListener {
            nbHidePillClicked.set(
                true
            )
        }

        // Monet Pill
        binding.nbMonetPill.setSwitchChangeListener { _: CompoundButton?, isSwitchChecked: Boolean ->
            Handler(Looper.getMainLooper()).postDelayed(
                {
                    if (isSwitchChecked) {
                        enableOverlay("IconifyComponentNBMonetPill.overlay")
                        restartSystemUI()
                    } else {
                        OverlayUtils.disableOverlay("IconifyComponentNBMonetPill.overlay")
                        restartSystemUI()
                    }
                }, SWITCH_ANIMATION_DELAY
            )
        }

        // Hide Keyboard Buttons
        binding.nbHideKbButtons.setSwitchChangeListener { _: CompoundButton?, isSwitchChecked: Boolean ->
            Handler(Looper.getMainLooper()).postDelayed(
                {
                    if (isSwitchChecked) {
                        enableOverlay("IconifyComponentNBHideKBButton.overlay")
                    } else {
                        OverlayUtils.disableOverlay("IconifyComponentNBHideKBButton.overlay")
                    }
                }, SWITCH_ANIMATION_DELAY
            )
        }

        // Disable left gesture
        binding.nbDisableLeftGesture.setSwitchChangeListener { _: CompoundButton?, isSwitchChecked: Boolean ->
            Handler(Looper.getMainLooper()).postDelayed(
                {
                    if (isSwitchChecked) {
                        Shell.cmd(
                            "settings put secure back_gesture_inset_scale_left -1 &>/dev/null"
                        ).exec()
                    } else {
                        Shell.cmd(
                            "settings delete secure back_gesture_inset_scale_left &>/dev/null"
                        ).exec()
                    }
                }, SWITCH_ANIMATION_DELAY
            )
        }

        // Disable right gesture
        binding.nbDisableRightGesture.setSwitchChangeListener { _: CompoundButton?, isSwitchChecked: Boolean ->
            Handler(Looper.getMainLooper()).postDelayed(
                {
                    if (isSwitchChecked) {
                        Shell.cmd(
                            "settings put secure back_gesture_inset_scale_right -1 &>/dev/null"
                        ).exec()
                    } else {
                        Shell.cmd(
                            "settings delete secure back_gesture_inset_scale_right &>/dev/null"
                        ).exec()
                    }
                }, SWITCH_ANIMATION_DELAY
            )
        }

        // Pill shape
        binding.pillShape.pillShapeContainer.visibility =
            if (binding.nbFullscreen.isSwitchChecked || binding.nbHidePill.isSwitchChecked) {
                View.GONE
            } else {
                View.VISIBLE
            }

        // Pill width
        val finalPillWidth = intArrayOf(getInt(FABRICATED_PILL_WIDTH, 108))
        binding.pillShape.pillWidth.sliderValue = finalPillWidth[0]
        binding.pillShape.pillWidth.setOnSliderTouchListener(object :
            Slider.OnSliderTouchListener {
            override fun onStartTrackingTouch(slider: Slider) {}

            override fun onStopTrackingTouch(slider: Slider) {
                finalPillWidth[0] = slider.value.toInt()
            }
        })

        // Pill thickness
        val finalPillThickness = intArrayOf(getInt(FABRICATED_PILL_THICKNESS, 2))
        binding.pillShape.pillThickness.sliderValue = finalPillThickness[0]
        binding.pillShape.pillThickness.setOnSliderTouchListener(object :
            Slider.OnSliderTouchListener {
            override fun onStartTrackingTouch(slider: Slider) {}

            override fun onStopTrackingTouch(slider: Slider) {
                finalPillThickness[0] = slider.value.toInt()
            }
        })

        // Bottom space
        val finalBottomSpace = intArrayOf(getInt(FABRICATED_PILL_BOTTOM_SPACE, 6))
        binding.pillShape.pillBottomSpace.sliderValue = finalBottomSpace[0]
        binding.pillShape.pillBottomSpace.setOnSliderTouchListener(object :
            Slider.OnSliderTouchListener {
            override fun onStartTrackingTouch(slider: Slider) {}

            override fun onStopTrackingTouch(slider: Slider) {
                finalBottomSpace[0] = slider.value.toInt()
            }
        })

        // Apply button
        binding.pillShape.pillShapeApply.setOnClickListener {
            if (!hasStoragePermission()) {
                requestStoragePermission(requireContext())

                return@setOnClickListener
            }

            Handler(Looper.getMainLooper()).post {
                putBoolean(PILL_SHAPE_SWITCH, true)

                putInt(FABRICATED_PILL_WIDTH, finalPillWidth[0])
                putInt(FABRICATED_PILL_THICKNESS, finalPillThickness[0])
                putInt(FABRICATED_PILL_BOTTOM_SPACE, finalBottomSpace[0])

                buildOverlayWithResource(
                    ResourceEntry(
                        SYSTEMUI_PACKAGE,
                        "dimen",
                        "navigation_home_handle_width",
                        finalPillWidth[0].toString() + "dp"
                    ),
                    ResourceEntry(
                        SYSTEMUI_PACKAGE,
                        "dimen",
                        "navigation_handle_radius",
                        finalPillThickness[0].toString() + "dp"
                    ),
                    ResourceEntry(
                        SYSTEMUI_PACKAGE,
                        "dimen",
                        "navigation_handle_bottom",
                        finalBottomSpace[0].toString() + "dp"
                    )
                )

                binding.pillShape.pillShapeReset.visibility = View.VISIBLE

                Handler(Looper.getMainLooper()).postDelayed(
                    { restartSystemUI() },
                    SWITCH_ANIMATION_DELAY
                )
            }
        }

        // Reset button
        binding.pillShape.pillShapeReset.visibility =
            if (getBoolean(PILL_SHAPE_SWITCH)) {
                View.VISIBLE
            } else {
                View.GONE
            }
        binding.pillShape.pillShapeReset.setOnClickListener {
            if (!hasStoragePermission()) {
                requestStoragePermission(requireContext())

                return@setOnClickListener
            }

            Handler(Looper.getMainLooper()).post {
                putBoolean(PILL_SHAPE_SWITCH, false)

                removeResourceFromOverlay(
                    ResourceEntry(SYSTEMUI_PACKAGE, "dimen", "navigation_home_handle_width"),
                    ResourceEntry(SYSTEMUI_PACKAGE, "dimen", "navigation_handle_radius"),
                    ResourceEntry(SYSTEMUI_PACKAGE, "dimen", "navigation_handle_bottom")
                )

                binding.pillShape.pillShapeReset.visibility = View.GONE

                Handler(Looper.getMainLooper()).postDelayed(
                    { restartSystemUI() },
                    SWITCH_ANIMATION_DELAY
                )
            }
        }
        return view
    }

    private fun initializeLeftGestureSwitch(): Boolean {
        return try {
            Shell.cmd(
                "settings get secure back_gesture_inset_scale_left"
            ).exec().out[0].toInt() == -1
        } catch (ignored: Exception) {
            false
        }
    }

    private fun initializeRightGestureSwitch(): Boolean {
        return try {
            Shell.cmd(
                "settings get secure back_gesture_inset_scale_right"
            ).exec().out[0].toInt() == -1
        } catch (ignored: Exception) {
            false
        }
    }

    private fun disableOthers(identifier: String) {
        if (identifier != NAVBAR_FULL_SCREEN) {
            putBoolean(NAVBAR_FULL_SCREEN, false)
            binding.nbFullscreen.isSwitchChecked = false
        }

        if (identifier != NAVBAR_IMMERSIVE_V1) {
            putBoolean(NAVBAR_IMMERSIVE_V1, false)
            binding.nbImmersive.isSwitchChecked = false
        }

        if (identifier != NAVBAR_IMMERSIVE_V2) {
            putBoolean(NAVBAR_IMMERSIVE_V2, false)
            binding.nbImmersiveV2.isSwitchChecked = false
        }

        if (identifier != NAVBAR_IMMERSIVE_V3) {
            putBoolean(NAVBAR_IMMERSIVE_V3, false)
            binding.nbImmersiveV3.isSwitchChecked = false
        }
    }

    private fun handleFullScreen(enable: Boolean) {
        putBoolean(NAVBAR_FULL_SCREEN, enable)

        val gcamLagFix = getBoolean(NAVBAR_GCAM_LAG_FIX)

        if (enable) {
            val resource = if (gcamLagFix) "1dp" else "0dp"

            buildOverlayWithResource(
                ResourceEntry(FRAMEWORK_PACKAGE, "bool", "config_imeDrawsImeNavBar", "false"),
                ResourceEntry(FRAMEWORK_PACKAGE, "dimen", "navigation_bar_height", resource),
                ResourceEntry(FRAMEWORK_PACKAGE, "dimen", "navigation_bar_frame_height", resource)
            )
        } else {
            removeResourceFromOverlay(
                ResourceEntry(FRAMEWORK_PACKAGE, "bool", "config_imeDrawsImeNavBar"),
                ResourceEntry(FRAMEWORK_PACKAGE, "dimen", "navigation_bar_height"),
                ResourceEntry(FRAMEWORK_PACKAGE, "dimen", "navigation_bar_frame_height")
            )
        }
    }

    private fun handleImmersive(enable: Boolean, @IntRange(from = 1, to = 3) version: Int) {
        when (version) {
            1 -> {
                putBoolean(NAVBAR_IMMERSIVE_V1, enable)
            }

            2 -> {
                putBoolean(NAVBAR_IMMERSIVE_V2, enable)
            }

            3 -> {
                putBoolean(NAVBAR_IMMERSIVE_V3, enable)
            }
        }

        val gcamLagFix = getBoolean(NAVBAR_GCAM_LAG_FIX)

        if (enable) {
            val resource = if (gcamLagFix) "1dp" else "0dp"
            val frameResource = if (version == 1) "48dp" else if (version == 2) "26dp" else "16dp"

            buildOverlayWithResource(
                ResourceEntry(
                    FRAMEWORK_PACKAGE,
                    "bool",
                    "config_imeDrawsImeNavBar",
                    "false"
                ),
                ResourceEntry(
                    FRAMEWORK_PACKAGE,
                    "dimen",
                    "navigation_bar_height",
                    resource
                ),
                ResourceEntry(
                    FRAMEWORK_PACKAGE,
                    "dimen",
                    "navigation_bar_frame_height",
                    frameResource
                )
            )
        } else {
            removeResourceFromOverlay(
                ResourceEntry(
                    FRAMEWORK_PACKAGE,
                    "bool",
                    "config_imeDrawsImeNavBar"
                ),
                ResourceEntry(
                    FRAMEWORK_PACKAGE,
                    "dimen",
                    "navigation_bar_height"
                ),
                ResourceEntry(
                    FRAMEWORK_PACKAGE,
                    "dimen",
                    "navigation_bar_frame_height"
                )
            )
        }
    }

    private fun handleLowSensitivity(enable: Boolean) {
        putBoolean(NAVBAR_LOW_SENS, enable)
        if (enable) {
            buildOverlayWithResource(
                ResourceEntry(
                    FRAMEWORK_PACKAGE,
                    "dimen",
                    "navigation_bar_gesture_height",
                    "18dp"
                )
            )
        } else {
            removeResourceFromOverlay(
                ResourceEntry(
                    FRAMEWORK_PACKAGE,
                    "dimen",
                    "navigation_bar_gesture_height"
                )
            )
        }
    }

    private fun handleHidePill(enable: Boolean) {
        putBoolean(NAVBAR_HIDE_PILL, enable)
        if (enable) {
            buildOverlayWithResource(
                ResourceEntry(
                    SYSTEMUI_PACKAGE,
                    "dimen",
                    "navigation_handle_radius",
                    "0dp"
                ),
                ResourceEntry(
                    SYSTEMUI_PACKAGE,
                    "dimen",
                    "navigation_home_handle_width",
                    "0dp"
                )
            )
        } else {
            removeResourceFromOverlay(
                ResourceEntry(
                    SYSTEMUI_PACKAGE,
                    "dimen",
                    "navigation_handle_radius"
                ),
                ResourceEntry(
                    SYSTEMUI_PACKAGE,
                    "dimen",
                    "navigation_home_handle_width"
                )
            )
        }
    }
}